src/Language/Spectacle/Syntax/Modal/Preterm.hs

```diff
@@ -0,0 +1,316 @@
+
+-- | 'Preterm' is the raw underlying syntax tree representing a temporal formula modulo the continuation or
+-- 'Data.Functor.Loom' information. A tree of 'Preterm's can be built by temporarily dispatching the 'Modal' and 'Logic'
+-- effects with 'materialize' and converted back into their effectful representation with 'abstract', which forms an
+-- isomorphism:
+--
+-- @
+--                                                      'materialize'
+--                                   ───────────────────────────────────────────────>
+-- Members '[Modal, Logic] effs =>                                                      Preterm Bool
+--   Lang ctx effs Bool              <───────────────────────────────────────────────
+--                                                       'abstract'
+-- @
+--
+-- 'Preterm's are rewritten in the normalization process rather than the effects directly since interposing effect
+-- translations directly would be extremely complicated for the set of tautologies used to normalize temporal formula.
+--
+-- @since 0.1.0.0
+data Preterm a where
+  PreConst :: a -> Preterm a
+  PreConjunct :: Preterm a -> Preterm a -> Preterm a
+  PreDisjunct :: Preterm a -> Preterm a -> Preterm a
+  PreImplies :: Preterm a -> Preterm a -> Preterm a
+  PreNotImplies :: Preterm a -> Preterm a -> Preterm a
+  PreComplement :: Preterm a -> Preterm a
+  PreAlways :: Preterm a -> Preterm a
+  PreUpUntil :: Preterm a -> Preterm a -> Preterm a
+  deriving (Functor, Eq, Show)
+
+isComplement :: Preterm a -> Bool
+isComplement PreComplement {} = True
+isComplement _ = False
+{-# INLINE CONLIKE isComplement #-}
+
+-- | 'materialize' takes a 'Lang' with 'Modal' and 'Logic' effects and converts their representation to 'Preterm's.
+--
+-- @since 0.1.0.0
+materialize :: Lang ctx (Modal ': Logic ': effs) a -> Lang ctx effs (Preterm a)
+materialize = \case
+  Pure x -> pure (PreConst x)
+  Op op k -> case decomposeOp op of
+    Left op' -> case decomposeOp op' of
+      Left other -> Op other (materialize . k)
+      Right (Logic b) -> absurd b
+    Right (Modal b) -> absurd b
+  Scoped scoped loom -> case decomposeS scoped of
+    Left scoped' -> case decomposeS scoped' of
+      Left other -> Scoped other loom'
+      Right eff
+        | Conjunct lhs rhs <- eff ->
+          runLoom loom' lhs >>= \case
+            PreComplement lhs' -> do
+              rhs' <- runLoom loom' rhs
+              if isComplement rhs'
+                then return (PreConjunct lhs' rhs')
+                else return (PreImplies lhs' rhs')
+            lhs' -> do
+              rhs' <- runLoom loom' rhs
+              return (PreConjunct lhs' rhs')
+        | Disjunct lhs rhs <- eff -> do
+          lhs' <- runLoom loom' lhs
+          rhs' <- runLoom loom' rhs
+          return (PreDisjunct lhs' rhs')
+        | Complement expr <- eff -> do
+          expr' <- runLoom loom' expr
+          return (PreComplement expr')
```

Comment:
I'm curious why you used guards instead of a second level of `case ... of` here to match on `eff`? Does GHC still do exhaustiveness checking with the guards?

To be clear, I'm not suggesting you need to change it to `case ... of` if you prefer it this way, I'm just curious if you had a specific reason.

Reply:
..so this is slightly wrong since that case expression should be matching in disjunct rather than conjunct.

The second case is /should be/ looking for a pattern of the form `PreDisjunct (PreComplement p) q` where q doesn't have `PreComplement` immediately nested inside of it. I do this here so I can take expressions of the form `¬p ∨ q` to be `p ⇒ q` which makes some things easier in the model checker down the line.

Something like `¬p ∨ ¬q` overlaps the match here but is not useful to rewrite this as `p ⇒ ¬ q` even though you could if you wanted to.